go/apps/mixnet/router.go

```diff
@@ -84,8 +94,53 @@ func (hp RouterContext) AcceptProxy() (net.Conn, error) {
 }
 
 // Close releases any resources held by the hosted program.
-func (hp RouterContext) Close() {
+func (hp *RouterContext) Close() {
 	if hp.proxyListener != nil {
 		hp.proxyListener.Close()
 	}
 }
+
+// HandleProxy reads a directive or a message from a proxy.
+func (hp *RouterContext) HandleProxy(c net.Conn) error {
+	var err error
+	cell := make([]byte, CellBytes)
+	if _, err = c.Read(cell); err != nil && err != io.EOF {
+		return err
+	}
+
+	// The first byte signals either a message or a directive to the router.
+	if cell[0] == msgCell {
+
+		// The first eight bytes of the first cell encode the message length.
+		// TODO(cjpatton) How to deal with endianness discrepancies?
+		msgBytes := int(binary.BigEndian.Uint64(cell[1:9]))
+		hp.msgBuffer = make([]byte, msgBytes)
+		bytes := copy(hp.msgBuffer, unpadCell(cell[9:]))
+
+		// While the connection is open and the message is incomplete, read
+		// the next cell.
+		for err != io.EOF && bytes < msgBytes {
+			if _, err = c.Read(cell); err != nil && err != io.EOF {
+				return err
+			}
+			bytes += copy(hp.msgBuffer[bytes:], unpadCell(cell))
+		}
+
+	} else if cell[0] == dirCell {
+		var d Directive
+		if err := proto.Unmarshal(unpadCell(cell)[1:], &d); err != nil {
+			return err
+		}
+
+		if *d.Type == DirectiveType_CREATE_CIRCUIT {
+			if len(d.Addrs) == 0 {
+				return errBadDirective
+			} else if len(d.Addrs) > 1 {
```

Comment:
Please pull this if statement / error return out to its own line instead of putting it in the else.